pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -18,60 +18,41 @@ package plugin
 
 import (
 	"context"
-	"errors"
 	"fmt"
-	"io"
-	"net"
 	"time"
 
 	"google.golang.org/grpc"
 	grpccodes "google.golang.org/grpc/codes"
-	"google.golang.org/grpc/credentials/insecure"
 	grpcstatus "google.golang.org/grpc/status"
 	"k8s.io/klog/v2"
 
 	drapbv1alpha2 "k8s.io/kubelet/pkg/apis/dra/v1alpha2"
-	drapb "k8s.io/kubelet/pkg/apis/dra/v1alpha3"
+	drapbv1alpha3 "k8s.io/kubelet/pkg/apis/dra/v1alpha3"
 )
 
 const PluginClientTimeout = 10 * time.Second
 
-// Strongly typed address.
-type draAddr string
-
-// draPluginClient encapsulates all dra plugin methods.
-type draPluginClient struct {
-	pluginName        string
-	addr              draAddr
-	nodeClientCreator nodeClientCreator
-}
-
-var _ drapb.NodeClient = &draPluginClient{}
-
-type nodeClientCreator func(addr draAddr) (
-	nodeClient drapb.NodeClient,
-	nodeClientOld drapbv1alpha2.NodeClient,
-	closer io.Closer,
-	err error,
-)
-
-// newNodeClient creates a new NodeClient with the internally used gRPC
-// connection set up. It also returns a closer which must be called to close
-// the gRPC connection when the NodeClient is not used anymore.
-// This is the default implementation for the nodeClientCreator, used in
-// newDRAPluginClient.
-func newNodeClient(addr draAddr) (nodeClient drapb.NodeClient, nodeClientOld drapbv1alpha2.NodeClient, closer io.Closer, err error) {
-	var conn *grpc.ClientConn
-
-	conn, err = newGrpcConn(addr)
+func detectVersion(conn *grpc.ClientConn) string {
+	client := drapbv1alpha3.NewNodeClient(conn)
+	_, err := client.NodePrepareResources(context.TODO(), &drapbv1alpha3.NodePrepareResourcesRequest{})
 	if err != nil {
-		return nil, nil, nil, err
+		status, _ := grpcstatus.FromError(err)
+		if status.Code() == grpccodes.Unimplemented {
+			return "v1alpha2"
+		}
 	}
 
-	return drapb.NewNodeClient(conn), drapbv1alpha2.NewNodeClient(conn), conn, nil
+	return "v1alpha3"
+}
```

Comment:
I guess the weird thing about this is that if there is an error that is *not* `grpccodes.Unimplemented`, we just assume it is version `v1alpha3`. I guess that is fine for now as we expect an error from the input you gave so there is no real way around this. Probably worth a comment though.

Reply:
Yes I do agree, this is not clean as I like but I spent a bit of time trying to find a way around w/o success.
One thing we could look in the future is maybe trying to retrieve metadata from gRPC server, we might be able to determine if it supports the alpha3 API